### PR TITLE
Avoid using invalid matches

### PR DESCRIPTION
**Pull request by HannesSommer**
This is a replacement of #229  (I've accidentally pushed to master effectively merging it; and undoing that does not reopen the PR..)

(addressing #228)

### CODE REVIEW

File: pointmatcher/InspectorsImpl.cpp [hunk 1]
```diff
@@ -320,25 +320,42 @@ void InspectorsImpl<T>::AbstractVTKInspector::dumpDataLinks(
 	}
 	const int knn = matches.ids.rows();
 	
-	stream << "LINES " << readingPtCount*knn << " "  << readingPtCount*knn * 3 << "\n";
+	size_t matchCount = readingPtCount*knn;
+	for(int k = 0; k < knn; ++k)
+	{
+		for (int i = 0; i < readingPtCount; ++i)
+		{
+			if(matches.ids(k, i) == Matches::InvalidId){
```
Comment (stephanemagnenat) [inline pointmatcher/InspectorsImpl.cpp]: This is a detail but the style (space after `if` and `{` at the end of the line) does not match the rest of the code.

**Comment by HannesSommer** [comment]
@bochen87 please use this state of the branch for testing.

I had to do quite some changes fixing several places where invalid matches were actually used.

Could somebody please have a closer look on the changes?

**Comment by bochen87** [comment]
@HannesSommer 
Couldn't compile, compile errors:
```

Errors     << libpointmatcher:make /opt/enway/ws/logs/libpointmatcher/build.make.002.log                                                                                                                   
libpointmatcher/examples/compute_overlap.cpp: In function ‘int main(int, char**)’:
libpointmatcher/examples/compute_overlap.cpp:223:40: error: ‘InvalidDist’ is not a member of ‘Matches {aka PointMatcher<float>::Matches}’
       if (targetMatches.dists(k, i) != Matches::InvalidDist)
                                        ^
make[2]: *** [examples/CMakeFiles/compute_overlap.dir/compute_overlap.cpp.o] Error 1
make[1]: *** [examples/CMakeFiles/compute_overlap.dir/all] Error 2

```

**Comment by bochen87** [comment]
Sorry, wrong warning - I cleaned up build and everything and built from scratch, worked now

**Comment by HannesSommer** [comment]
This is probably a local issue. Are you compiling it with catkin? In any case you: you are using outdated headers that probably were installed through `make install`. Try to find an delete them. In case of catkin delete `devel/includes/libpointmatcher`. Otherwise you have to track which header is actually used OR try `make uninstall` in pointmatchers build folder.

**Comment by bochen87** [comment]
👍 
wow, thanks for the really fast replies! yeah, somehow figured that out as well, InvalidDist was clearly defined in the header, so i cleaned up build folder and it worked perfectly, now on to testing!

**Comment by bochen87** [comment]
I've done some stability testing, let it run for 30+ mins - seems to work.
Will do some more test cases tomorrow, probably safe to merge after tests of tomorrow.

**Comment by bochen87** [comment]
@HannesSommer Did some more dynamic obstacle tests and it seems to be working stable!

**Comment by HannesSommer** [comment]
Nice, thanks for testing and reporting. 

Does somebody have time to go over the changes before we merge? 

In particular I'm not 100% sure about my decision to set the weights (OutlierFilters) to zero for all invalid matches as in https://github.com/ethz-asl/libpointmatcher/pull/230/files#diff-d26b55af8634c8550d46bdc8c70e2d91R255 .

**Comment by PhiBabin** [comment]
The other outlier filters (VarTrim, Trim and Welsch) set a zero weight to matches that should be ignore by the error minimizer.  It makes sense that an invalid match should not contribute to the error minimization.

File: pointmatcher/InspectorsImpl.cpp [hunk 2]
```diff
@@ -320,25 +320,42 @@ void InspectorsImpl<T>::AbstractVTKInspector::dumpDataLinks(
...
+		for (int i = 0; i < readingPtCount; ++i)
+		{
+			if (matches.ids(k, i) == Matches::InvalidId){
+				matchCount --;
+			}
+		}
+	}
+
+	stream << "LINES " << matchCount << " "  << matchCount * 3 << "\n";
 	//int j = 0;
```
Comment (PhiBabin) [inline pointmatcher/InspectorsImpl.cpp]: Dead code

File: pointmatcher/InspectorsImpl.cpp [hunk 3]
```diff
@@ -320,25 +320,42 @@ void InspectorsImpl<T>::AbstractVTKInspector::dumpDataLinks(
...
+				stream << "2 " << refPtCount + i << " " << id << "\n";
+			}
 		}
 	}
 
-	stream << "CELL_DATA " << readingPtCount*knn << "\n";
+	stream << "CELL_DATA " << matchCount << "\n";
 	stream << "SCALARS outlier float 1\n";
 	stream << "LOOKUP_TABLE default\n";
 	//stream << "LOOKUP_TABLE alphaOutlier\n";
```
Comment (PhiBabin) [inline pointmatcher/InspectorsImpl.cpp]: Dead code

**Comment by pomerlef** [comment]
I've aksed @PhiBabin to take a look. He's working on outlier rejection on our side.

File: pointmatcher/InspectorsImpl.cpp (hunk 2 above)
Comment (HannesSommer) [reply pointmatcher/InspectorsImpl.cpp]: I didn't touch these old dead lines and I would leave general cleaning up to cleanup PRs (makes the reviewing easier)

File: pointmatcher/InspectorsImpl.cpp (hunk 3 above)
Comment (HannesSommer) [reply pointmatcher/InspectorsImpl.cpp]: Same here.

**Comment by HannesSommer** [comment]
Thanks @PhiBabin  and @pomerlef  !

Then we can merge I assume?

**Comment by pomerlef** [comment]
All good for me. Thanks @HannesSommer for taking care of that.

**Pull request merged**